Rework of CBS to use dictionaries

PeterKraus opened the pull request:
## Description
The general idea is to allow more flexible definition of CBS functions, and clean up the `cbs()` function a little. The goal is to allow further corrections to the energy into CBS, eg differences of basis sets (eg. `\Delta E_{diff} = E(aug-cc-pvtz) - E(cc-pvtz)`) or frozen core corrections (`\Delta E_{ae} = E(freeze_core True) - E(freeze_core False)` or DKH ...

Added `frequency` to cbs-compatible calls.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Simple, dict-based interface to CBS.
- [x] Handling of an arbitrary number of correction functions.
- [x] Support for passing of arbitrary options (such as `freeze_core`). 
- [x] Fix `pywrap-cbs1` which fails on `cbs_number` checks.
- [x] Documentation.
- [x] Frequency calculations via cbs are now possible.

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

PeterKraus commented:
I've tried to re-write some of the cbs docs more thoroughly, however my sphinx-fu is not up to scratch. Especially the bits in `driver_cbs.py` under Combined interface, I am not sure what the best way of markup is - the current one is quite ugly.

PeterKraus commented:
Also, as a to-do for a different PR, it might be a good idea to unify the whole metadata structure with what's currently in `NEED`, `GRAND_NEED` and `MODELCHEM`, so that it's all in one place and not copied across 3 dictionaries.

Another wished-for feature is allowing `freq()` calls into cbs, superficially it doesn't look that hard, at least for methods where we can compute a cbs-`hessian()` already... But as I said above, it's on my radar, but not for this PR.

dgasmith commented:
Can you rebase this with master so that appveyor will pass.

PeterKraus commented:
So I've added the option to be able to run frequency calculations via the cbs interface, both using the `"method/basis"` and `frequency(cbs, ...)` syntax. I've added the tests for this too - I wanted to get a reference value for the `HF/cc-pV[DT]Z` extrapolation from other codes (thanks @hokru), but we failed to get it run correctly. When compared to large basis sets, one gets these results:

Freq 1 | Freq 2 | Freq 3 | zeta | RMS
-- | -- | -- | -- | -- 
1809.246 | 3923.151 | 4020.177 | 2 |    174.1217 
1753.040 | 4126.975 | 4226.872 | 3 |     4.1162
1750.381 | 4130.709 | 4230.122 | 4 |     1.4733
1748.147 | 4130.445 | 4230.985 | 5 |     0.2300
1748.003 | 4130.108 | 4230.827 | 6 |     
1747.741 | 4130.226 | 4230.575 | [dt] |    0.2207
1749.876 | 4131.278 | 4230.490 | [tq] |    1.2899
1747.595 | 4130.375 | 4231.193 | [q5] |    0.3520

I think the convergence is reasonable - it's clear the results (even with HF/[dt]) are not complete junk.

PeterKraus commented on psi4/driver/driver.py:
```diff
@@ -1746,18 +1747,19 @@ def frequency(name, **kwargs):
+    #     return driver_cbs._cbs_gufunc(frequency, name, ptype='frequency', **kwargs)
+    # 
+    # if kwargs.get('bsse_type', None) is not None:
+    #     raise ValdiationError("Frequency: Does not currently support 'bsse_type' arguements")
```
I've kept this piece of code for review, but it can be deleted later.

PeterKraus commented on psi4/driver/driver_cbs.py:
```diff
@@ -1681,6 +1488,7 @@ def cbs(func, label, **kwargs):
             finalquantity.print_out()
     elif ptype == 'hessian':
         finalquantity = finalhessian
+        wfn.set_gradient(finalgradient)
```
If there is a good reason for not doing this with all `frequency(cbs, ...)` calculations, let me know - otherwise the frequency calculations fail (as the gradient is not set properly).

loriab replied:
good addition

PeterKraus commented on tests/cbs-xtpl-freq/input.dat:
```diff
@@ -3,42 +3,61 @@ import numpy as np
   H 1 0.96 2 104.5
 }
 
+molecule h2o_tz {
```
These two were optimized using `g_convergence gau_verytight`

PeterKraus commented:
So the windows test fails. I am assuming it's because it's using (implicit) `dertype=2` for the scf frequency - is this a known fail on Windows? @raimis, do you have any idea?

raimis commented:
Also `cbs-parser` and `cbs-xtpl-func` are still failing, but I haven't yet tried to debug them. So, at the moment, no idea what might be wrong.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -49,20 +49,31 @@
+    
+    Parameters
+    ----------
+    basisstring: string
```
space before `:`. sphinx is very picky

loriab replied:
also, probably `str` rather than `string`

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -703,6 +893,149 @@ def return_energy_components():
+                metadata[-1]["treatment"] = item.get("treatment", "scf" if len(metadata) == 1 else "corl")
+                metadata[-1]["scheme"] = item.get("scheme", _get_default_xtpl(len(metadata[-1]["basis"][1]),
+                                                                                 metadata[-1]["treatment"])) 
+                if len(metadata) > 1:
```
doesn't this only kick in for deltas at `len(metadata) > 2:`?

PeterKraus replied:
No, the treatment of correlation (index == 1) and following delta-correlations (index >1) is essentially identical.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -703,6 +893,149 @@ def return_energy_components():
+                    metadata[-1]["wfn_lo"] = item.get("wfn_lo", metadata[-2].get("wfn"))
+                    metadata[-1]["basis_lo"] = _expand_bracketed_basis(item.get("basis_lo", 
+                                                                       item.get("basis")).lower())
+                    if len(metadata[-1]["basis"][0]) != len(metadata[-1]["basis_lo"][0]):
```
I wonder if it'd be a bit cleaner to have a `trial_stage = {}` at the start, fill that instead of `metadata[-1]`, then `metadata.append(trial_stage)` at the end if all successful. Open for debate.

PeterKraus replied:
Fine by me.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -703,6 +893,149 @@ def return_energy_components():
+                    elif len(metadata) == 2:
+                        metadata[-1]["stage"] = "corl"
+                    else:
+                        metadata[-1]["stage"] = "delta{0:d}".format(len(metadata) - 2)
```
If you're doing the `0` in `delta{0:d}` in order to specify the item in the format string to format, that was only necessary in py26 or if you want to go out-of-order. 

You may have a better reason -- I'm not totally up on format string syntax.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -703,6 +893,149 @@ def return_energy_components():
+                    else:
+                        metadata[-1]["stage"] = "delta{0:d}".format(len(metadata) - 2)
+            metadata[-1]["alpha"] = item.get("alpha", None)
+            metadata[-1]["options"] = item.get("options", False)
```
This'll be great. I've long considered lack of stage-directed options as one of cbs()'s greatest weaknesses.

PeterKraus replied:
Yeah - it's what motivated me to do this in the first place.

JonathonMisiewicz commented on psi4/driver/driver_cbs.py:
```diff
@@ -1076,23 +1365,21 @@ def cbs(func, label, **kwargs):
-        'delta2': kwargs.get('cbs_delta2_alpha', kwargs.get('delta2_alpha', None)),
-    }
 
     # Establish function to call (only energy makes sense for cbs)
```
This comment, and the error message two lines below, are inaccurate now.

JonathonMisiewicz commented:
Two comments about lines that weren't changed:
https://github.com/psi4/psi4/blob/a658c8e50e6d5d7bde2b8058b33e78f98f800db3/psi4/driver/driver_cbs.py#L1910-L1912
We shouldn't be dropping out for frequencies anymore.

https://github.com/psi4/psi4/blob/a658c8e50e6d5d7bde2b8058b33e78f98f800db3/psi4/driver/driver_cbs.py#L1940-L1944
It would be good to put a comment here explicitly saying that `_parse_cbs_gufunc_string` should enforce that `len(method_list)` can only be 1 or 2. The logic confused me for a few minutes, on first read.

PeterKraus commented on psi4/driver/driver_cbs.py:
```diff
@@ -703,6 +893,149 @@ def return_energy_components():
+                    raise ValidationError("Basis set has to be supplied for every stage.")
+                else:
+                    metadata[-1]["basis"] = _expand_bracketed_basis(item.get("basis").lower())
+                metadata[-1]["treatment"] = item.get("treatment", "scf" if len(metadata) == 1 else "corl")
```
I'm not sure I understand. The old behaviour of just specifying "corl_wfn" and higher, without specifying "scf_wfn" still works. For "cbs_metadata", the first item in the array is by default a "scf" stage, which means a scf extrapolation method will be used (as opposed to corl extrapolation) - it can (as before) be an extrapolated, or highest_1 stage, but is not restricted to "scf"/"hf"/"c4-" anymore. The extrapolation (scf vs corl) is overridable by providing the "treatment" key, the name of the stage is overridable by the "stage" key.

loriab replied:
But is it correct that `energy(cbs, cbs_metadata=[{'wfn': 'mp2', 'basis': 'cc-pv[dt]z'}])` is no longer valid b/c no scf stage? What did you have in mind as something ppl would want other than HF (in various flavors and sources)?

PeterKraus replied:
So there are a few options:
```
[1] energy("mp2/cc-pv[dt]z+d:ccsd(t)/cc-pvdz")
[2] energy(cbs, corl_wfn="mp2", corl_basis="cc-pv[dt]z", 
                delta_wfn="ccsd(t)", delta_basis="cc-pvdz")
[3] energy(cbs, cbs_metadata=[{"wfn": "hf", "basis": "cc-pvtz"},
                              {"wfn": "mp2", "basis": "cc-pv[dt]z"},
                              {"wfn": "ccsd(t)", "basis": "cc-pvdz"}])
E = hf/cc-pvtz + (mp2 - hf)/cc-pv[dt]z + (ccsd(t) - mp2)/cc-pvtz
E = -76.35375057

[4] energy(cbs, cbs_metadata=[{"wfn": "mp2", "basis": "cc-pv[dt]z"},
                              {"wfn": "ccsd(t)", "basis": "cc-pvdz"}])
E = mp2/cc-pv[dt]z + (ccsd(t) - mp2)/cc-pvdz
E = -76.35033786

[5] energy(cbs, cbs_metadata=[{"wfn": "mp2", "basis": "cc-pv[dt]z", "treatment": "corl"},
                              {"wfn": "ccsd(t)", "basis": "cc-pvdz"}])
E = mp2/cc-pv[dt]z + (ccsd(t) - mp2)/cc-pvdz
E = -76.36620899

```
The first three calls are equivalent and are the default behaviour. The last two differ from one another by calling "scf_xtpl_helgaker_2" in [4] and "corl_xtpl_helgaker_2" in [5] directly over the mp2 energies.

loriab replied:
I'm separately confused by your `[1--5]`. By my common practice (and hence cbs's), `mp2/cc-pv[dt]z == hf/cc-pvtz + (mp2 - hf)/cc-pv[dt]z` with the last term using scf_xtpl_helgaker_2. So `[1--5]` should all be the same. Are you reasoning differently?

If `[4]` is, in fact, the same as `[1--3]`, then that resolves my original concern. That is, is it necessary to explicitly have SCF as the first item in the `cbs_metadata` list? Apparently not, and all's well.

If it clarifies anything, I don't think one should ever apply a correlation energy extrapolation equation to a total energy, unless compensating by subtracting out the reference energy subjected to the same (unnatural b/c eqn designed for correlation) equation is the next step.

PeterKraus replied:
Yes, currently `[3]` and `[4]` are different, as in  `[1,2,3]` the energy is taken as:
```
  -------------------------------------------------------------------------------------
      Stage      Method / Basis             Wt      Energy [Eh]   Scheme
  -------------------------------------------------------------------------------------
        scf          hf / cc-pvtz            1     -76.05098620   xtpl_highest_1
       corl         mp2 / cc-pv[dt]z         1     -76.35320638   corl_xtpl_helgaker_2
       corl          hf / cc-pv[dt]z        -1     -76.06344462   corl_xtpl_helgaker_2
     delta1     ccsd(t) / cc-pvdz            1     -76.23906022   xtpl_highest_1
     delta1         mp2 / cc-pvdz           -1     -76.22605761   xtpl_highest_1
  -------------------------------------------------------------------------------------
```
In the fourth case it is:
```
  ------------------------------------------------------------------------------------
      Stage     Method / Basis             Wt      Energy [Eh]   Scheme
  ------------------------------------------------------------------------------------
        scf        mp2 / cc-pv[dt]z         1     -76.33733525   scf_xtpl_helgaker_2
       corl    ccsd(t) / cc-pvdz            1     -76.23906022   xtpl_highest_1
       corl        mp2 / cc-pvdz           -1     -76.22605761   xtpl_highest_1
  ------------------------------------------------------------------------------------
```
and in the fifth case:
```
  ------------------------------------------------------------------------------------
      Stage     Method / Basis             Wt      Energy [Eh]   Scheme
  ------------------------------------------------------------------------------------
        scf        mp2 / cc-pv[dt]z         1     -76.35320638   *corl_xtpl_helgaker_2*
       corl    ccsd(t) / cc-pvdz            1     -76.23906022   xtpl_highest_1
       corl        mp2 / cc-pvdz           -1     -76.22605761   xtpl_highest_1
  ------------------------------------------------------------------------------------
```
If you think allowing `[4,5]` is not a good idea, I can validate the `wfn` key of the first item and raise a validation error if it's not one of `HF/SCF/C4-HF/C4-SCF`.

loriab replied:
I _do_ think allowing `[4,5]` is a Bad Idea. Though I'd relent if everyone disagrees.

Validating `wfn` of the first item if not `HF/C4-HF` (fine by me if the vague `SCF` gets dropped) is a good plan. My notion of a better plan would be if first item not `HF/C4-HF`, just create `{'wfn': "hf", "treatment": highest_1, 'basis': copy_corl_basis_str}` such that the first iteration through `cbs_metadata` creates _two_ entries, the scf and corl stages. Whichever you prefer.

PeterKraus commented:
> Two comments about lines that weren't changed:
> 
> [psi4/psi4/driver/driver_cbs.py](https://github.com/psi4/psi4/blob/a658c8e50e6d5d7bde2b8058b33e78f98f800db3/psi4/driver/driver_cbs.py#L1910-L1912)
> 
> Lines 1910 to 1912 in [a658c8e](/psi4/psi4/commit/a658c8e50e6d5d7bde2b8058b33e78f98f800db3)
>  # Drop out for props and freqs 
>  if ptype in ["properties", "frequency"]: 
>      raise ValidationError("%s: Cannot extrapolate or delta correct %s yet." % (ptype.title(), ptype)) 
> 
> We shouldn't be dropping out for frequencies anymore.

Fair point. I changed the raw 'frequency' function too, to be a true wrapper around 'hessian', that code is just a left-over. Changed.

> [psi4/psi4/driver/driver_cbs.py](https://github.com/psi4/psi4/blob/a658c8e50e6d5d7bde2b8058b33e78f98f800db3/psi4/driver/driver_cbs.py#L1940-L1944)
> 
> Lines 1940 to 1944 in [a658c8e](/psi4/psi4/commit/a658c8e50e6d5d7bde2b8058b33e78f98f800db3)
>  if len(method_list) > 1: 
>      cbs_kwargs['delta_wfn'] = method_list[1] 
>      cbs_kwargs['delta_basis'] = basis_list[1] 
>      if 'delta_scheme' in kwargs: 
>          cbs_kwargs['delta_scheme'] = kwargs['delta_scheme'] 
> 
> It would be good to put a comment here explicitly saying that `_parse_cbs_gufunc_string` should enforce that `len(method_list)` can only be 1 or 2. The logic confused me for a few minutes, on first read.

I'll edit the docs. The idea is that one usually only specifies one "delta correction" in a string, ie:
```
mp2/cc-pv[tq]z+D:ccsd(t)/cc-pvtz
```
as opposed to trying to do two delta corrections in a "method/basis" syntax (which is not supported):
```
hf/cc-pv[q5]z+D:mp2/cc-pv[tq]z+D2:ccsd(t)/cc-pvtz
```
Folks that want to pile on the deltas can use the old interface (with scf, corl, delta and delta2 stages) or the cbs_metadata interface, which doesn't have a limit.

loriab commented:
I got that appveyor py27 error, too. Rebasing to bring in #1278 fixed it. No urgency, just when you're ready.

PeterKraus commented:
I'll rebase when I work on this next time. When you find some time, could you go through my comments in my own reviews and in yours? Also, I'm not sure what the conclusion of the "scf stage" debate above was - is it fine as it is, does it need more docs, or shall it be changed?

PeterKraus commented:
@psi4/editors: this one should be ready for a final review & merge.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -701,13 +936,170 @@ def return_energy_components():
+                stage = {"wfn": kwargs["{:s}_wfn".format(sn)], "basis": kwargs["{:s}_basis".format(sn)]}
+            # or we're at a scf stage which can be implied with a provided scf_basis
+            elif sn == "scf" and "{:s}_basis".format(sn) in kwargs:
+                stage = {"wfn": "hf", "basis": kwargs["{:s}_basis".format(sn)]}
```
Can this handle the internal setup for SCF stage right `energy(cbs, corl_wfn='mp2', corl_basis='cc-pv[dt]z')`? I'm not seeing it, but maybe I'm wrong. You've got the running code handy.

PeterKraus replied:
Yup, exactly the same behaviour as in 1.3a1:
[cbs_dict](https://github.com/psi4/psi4/files/2519138/post.txt), [1.3a1](https://github.com/psi4/psi4/files/2519139/pre.txt)

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -798,88 +1189,52 @@ def cbs(func, label, **kwargs):
-
-        Indicates the inferior energy method for which a fifth delta correction
-        to the correlation energy is to be obtained.
-
```
yay, deletions :-)

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -701,13 +936,170 @@ def return_energy_components():
+    # 2a) process required stage parameters and assign defaults
+        stage = {}
+        stage["wfn"] = item["wfn"].lower()
+        stage["basis"] = _expand_bracketed_basis(item["basis"].lower())
```
doesn't this need `molecule=molstr` as second arg to fn call? Can probably replace the str with dict nowadays.

PeterKraus replied:
It doesn't need it (as it builds the basis for H2 when no molecule is provided), but it might be safer to pass it on.

PeterKraus replied:
Fixed now.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -1839,6 +1651,9 @@ def _cbs_gufunc(func, total_method_name, **kwargs):
     cbs_kwargs['verbose'] = cbs_verbose
+    
+    if user_dertype != None:
+        cbs_kwargs['dertype'] = user_dertype
```
Not sure. If it was deliberate, recursiveness or sow/reap would be my suspects.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -1839,6 +1651,9 @@ def _cbs_gufunc(func, total_method_name, **kwargs):
+        cbs_kwargs['dertype'] = user_dertype
 
     # Find method and basis
     if method_list[0] in ['scf', 'hf', 'c4-scf', 'c4-hf']:
```
Sure, I think this is an excellent use for cbs_metadata,

PeterKraus commented on doc/sphinxman/source/cbs.rst:
```diff
@@ -181,3 +191,24 @@ existing examples are below.
+
+.. autofunction:: psi4.driver.driver_cbs._parse_cbs_gufunc_string
+
+.. autofunction:: psi4.driver.driver_cbs._cbs_gufunc
```
I wasn't sure what the policy is - whether we want to document each function, or just the user-facing ones. My thought was: if I spend the time writing the docs, I'll want them rendered in the manual. I can remove the section, if you prefer to keep the manual concise.

loriab replied:
I worry users will be confused and think they should make use of the internal fns. Go ahead and remove them, please. In a better version of the API docs for devs, they'll show up.

PeterKraus commented on psi4/driver/driver_cbs.py:
```diff
@@ -701,13 +936,170 @@ def return_energy_components():
+    -------
+    array
+        Array of dictionaries, with each item consisting of an extrapolation
+        stage. All validation takes place here.
```
I've replaced most occurrences of the word `array` with `list`.

PeterKraus commented on psi4/driver/driver_cbs.py:
```diff
@@ -701,13 +936,170 @@ def return_energy_components():
+    for item in cbs_metadata:
+        # 1a) all items must have wfn
+        if "wfn" not in item:
+            raise ValidationError("Stage {d} doesn't have defined level of theory!".format(cbs_metadata.index(item)))
```
I meant to use {:d}, actually. I'd prefer the expected type to be shown.

loriab commented on psi4/driver/driver_cbs.py:
```diff
@@ -701,13 +936,170 @@ def return_energy_components():
+            metadata.append(scf)
+    # 2c) keep processing current stage
+        stage["treatment"] = item.get("treatment", "scf" if len(metadata) == 0 else "corl")
+        stage["stage"] = item.get("stage", False)
```
I agree on `stage` -- it'll be nice to have variable labels. Is that in a test somewhere?

Not so sure about `treatment`. It's switching btwn `scf` and `corl`, right? I wouldn't have expected past code to handle a non-scf-or-corl value correctly. Great if it can now, just not seeing it from a quick search.

loriab replied:
Because of GitHub's infernal new "helpfullness" in hiding conversations, I can't find the `basis_lo` thread. I understand your use-case now and agree. And perhaps if there were several alternate formulas to be applied to `"AE - FC"` then that would constitute a new `treatment` class in your view? For `basis_lo`, since that's a new feature, I'd feel better if it had a test to make sure we don't break it.

PeterKraus replied:
In the past code, you could only ever pass in one `scf` stage. The new code, due to `_get_default_xtpl` requires each stage to have either `corl` or `scf` set using the `treatment` tag. One can, in principle, pass two `scf` stages using the new code. I can't think of why someone would want to do that, perhaps with the exception of relativistic corrections over scf values from extrapolated basis sets. It's a very niche use case (the xtpl formula for scf and corl for 1 basis set is the same - xtpl_highest_1). I can of course have the code ignore user input.

There is no test for `stage`; it is really just a print label - I didn't think it was necessary.

The `AE - FC` class is still correction (or extrapolation with 2 basis sets) of correlation; it doesn't require a new class of `treatment`.

I'll add a test for `basis_lo`.

loriab replied:
Thanks. I think the `basis_lo` is the important test. Throw in an alternate `stage` label, just so the line of code gets hit and we may notice if something breaks.

For `treatment`, pass `treatment='leeches'` and if a sensible error happens, I'm good with having it user-settable. If error is obscure, best to take that knob away from the user. (If calculation works, I'm backing away slowly.)

PeterKraus replied:
Leeches now throw a validation error ;)

PeterKraus commented:
@JonathonMisiewicz , as you also reviewed this PR, any comments?
